src/WinRT.Runtime/Projections/EventHandler.cs

```diff
@@ -88,130 +88,149 @@ public unsafe static bool InitRcwHelper(delegate*<IObjectReference, object, T, v
+                global::System.Threading.Interlocked.CompareExchange(ref _queryInterfaceCache, new ConcurrentDictionary<RuntimeTypeHandle, IObjectReference>(), null);
                 return _queryInterfaceCache;
             }
-            ConcurrentDictionary<RuntimeTypeHandle, IObjectReference> IWinRTObject.QueryInterfaceCache => _queryInterfaceCache ?? MakeQueryInterfaceCache();
-
-            private volatile ConcurrentDictionary<RuntimeTypeHandle, object> _additionalTypeData;
-            private ConcurrentDictionary<RuntimeTypeHandle, object> MakeAdditionalTypeData()
+            ConcurrentDictionary<RuntimeTypeHandle, IObjectReference> IWinRTObject.QueryInterfaceCache => _queryInterfaceCache ?? MakeQueryInterfaceCache();
+
+            private volatile ConcurrentDictionary<RuntimeTypeHandle, object> _additionalTypeData;
+            private ConcurrentDictionary<RuntimeTypeHandle, object> MakeAdditionalTypeData()
             {
-                global::System.Threading.Interlocked.CompareExchange(ref _additionalTypeData, new ConcurrentDictionary<RuntimeTypeHandle, object>(), null);
+                global::System.Threading.Interlocked.CompareExchange(ref _additionalTypeData, new ConcurrentDictionary<RuntimeTypeHandle, object>(), null);
                 return _additionalTypeData;
             }
-            ConcurrentDictionary<RuntimeTypeHandle, object> IWinRTObject.AdditionalTypeData => _additionalTypeData ?? MakeAdditionalTypeData();
-#endif
+            ConcurrentDictionary<RuntimeTypeHandle, object> IWinRTObject.AdditionalTypeData => _additionalTypeData ?? MakeAdditionalTypeData();
+#endif
 
-            public unsafe void Invoke(object sender, T args)
+            public unsafe void Invoke(object sender, T args)
             {
-#if NET
-                bool useDynamicInvoke = EventHandlerMethods<T>._Invoke == null && RuntimeFeature.IsDynamicCodeCompiled;
-#else
-                bool useDynamicInvoke = EventHandlerMethods<T>._Invoke == null;
-#endif
-                if (!useDynamicInvoke)
+#if NET
+                // Standalone path for NAOT to ensure the linker trims code as expected
+                if (!RuntimeFeature.IsDynamicCodeCompiled)
+                {
+                    EventHandlerMethods<T>._Invoke(_nativeDelegate, sender, args);
+
+                    return;
+                }
+#endif
+
+                if (EventHandlerMethods<T>._Invoke != null)
                 {
                     EventHandlerMethods<T>._Invoke(_nativeDelegate, sender, args);
-                }
+                }
                 else
                 {
+                    // Same as in the static constructor, we initialize the ABI delegate type manually here if needed
+                    global::System.Threading.Interlocked.CompareExchange(ref _abi_invoke_type, Projections.GetAbiDelegateType(typeof(void*), typeof(IntPtr), Marshaler<T>.AbiType, typeof(int)), null);

```

Comment:
Any opportunity here to avoid calling `Projections.GetAbiDelegateType` on every invoke when AOT or the new optimizer changes isn't being used?

Reply:
Ah, yeah good catch! Added a check in 99d4b2dd5d466baa547b8c018cc90bf54d754bb3 to avoid unnecessary calls 🙂